pkg/madmin/kms-commands.go

```diff
@@ -0,0 +1,66 @@
+/*
+ * MinIO Cloud Storage, (C) 2019 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package madmin
+
+import (
+	"encoding/json"
+	"net/http"
+	"net/url"
+)
+
+// GetKeyInfo requests information about the key referenced by keyID
+// from the KMS connected to a MinIO by performing a Admin-API request.
+// It basically hits the `/minio/admin/v1/kms/key/info` API endpoint.
+func (adm *AdminClient) GetKeyInfo(keyID string) (*KMSKeyInfoResponse, error) {
+	// GET /minio/admin/v1/kms/key/info?key-id=<keyID>
+	qv := url.Values{}
+	qv.Set("key-id", keyID)
+	reqData := requestData{
+		relPath:     "/v1/kms/key/info",
+		queryValues: qv,
+	}
+
+	resp, err := adm.executeMethod("GET", reqData)
+	defer closeResponse(resp)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, httpRespToErrorResponse(resp)
+	}
+	var keyInfo KMSKeyInfoResponse
+	if err = json.NewDecoder(resp.Body).Decode(&keyInfo); err != nil {
+		return nil, err
+	}
+	return &keyInfo, nil
+}
+
+// KMSKeyStatus is the status of a particular key at the KMS.
+// It indicates whether the key can be used by the MinIO server
+// for encryption as well as decryption.
+type KMSKeyStatus struct {
+	EncryptionErr string `json:"encryption-error,omitempty"` // An empty error == success
+	DecryptionErr string `json:"decryption-error,omitempty"` // An empty error == success
+}
+
+// KMSKeyInfoResponse is the response returned by a MinIO server
+// when the GetKeyInfo() call returns successfully. It contains
+// status information about one particular key at the KMS.
+type KMSKeyInfoResponse struct {
+	KeyID  string       `json:"key-id"`
+	Status KMSKeyStatus `json:"status"`
+}
```

Comment:
it is simpler to remove `type KMSKeyStatus` and have:

```
type KMSKeyInfoResponse struct {
	Err  error       `json:"err"`
}
```

and have `MarshalJSON()` and `UnmarshalJSON()` methods on this struct.

Err here can have encryption or decryption error. It is go

Reply:
There will be at least one more KMS operation (`UpdateKey`). Now with a single error we would need to rely on string parsing at the client. `Encryption` may succeed, `UpdateKey` may fail and `Decryption` may fail. How would you express such a situation in on error message?

Reply:
Also the `KeyID` field is need since the client may not send a key-ID and the server will use the default master key in this case. A `mc` client can learn the master key name this way.

Reply:
`type KMSKeyInfoResponse` I thought this was just for "key info" operation. You can rename it then to something more generic.

Reply:
Replaced key info with status. The `status` (as before) indicates whether the master key exists, is accessible by MinIO a.s.o. So, if the status request returns no error this means the configuration for that particular master key is correct.